packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,169 @@
+import assert from "assert";
+import { exists, getPositionKey } from "./utils";
+import { ethers } from "ethers";
+
+type Id = string | number;
+type HasId = {
+  id?: Id;
+};
+interface MakeId<T> {
+  (obj: T): string;
+}
+
+// Generic table with common functions, store is a map now, but we could replace this with any kind of
+// data store with some minor modifications, as long as it has, set, get, delete, has and values
+export function Table<T extends HasId>(config: { makeId: MakeId<T>; type: string }, store: Map<Id, T>) {
+  const { makeId, type } = config;
+  async function create(data: T) {
+    const id = makeId(data);
+    assert(!(await has(id)), `${type} exists`);
+    return set({ id, ...data });
+  }
+  async function set(data: T) {
+    assert(data.id, `${type} requires id`);
```

Comment:
a couple reasons, and this touches on why not make `id` required. The user doesn't always know the id of the data when sending it to the table. Its the tables responsibility with `makeId` to determine how to identify the data and insert this into the store.   Because of this, at the top of the Table definition I allow generic types without an id: ` Table<T extends HasId>` where `HasId` has an optional id, since the user might submit raw data without an `.id` on it.

I tried your suggestion though, and when I union `T` and `{id:Id}` it complains that its incompatible with ` Argument of type 'T & { id?: string | undefined; }' ` 

maybe theres a better way to do this though, i really am just mainly hacking